### PR TITLE
[ACQ-554] Style filter dropdowns on /curriculum

### PR DESCRIPTION
**Pull request by levadadenys**
### [ACQ-554] Style filter dropdowns on /curriculum

* styled filter dropdowns on /curriculum
* some refactor
* some cleanup

Before:
Знімок екрана 2023-05-26 о 17 23 05
Знімок екрана 2023-05-26 о 17 23 11

After:
Знімок екрана 2023-05-26 о 17 41 02
Знімок екрана 2023-05-26 о 17 41 09

## Links
- jira ticket: [ACQ-554](https://codedotorg.atlassian.net/browse/ACQ-554?atlOrigin=eyJpIjoiMTI3ZTRhMzg2MmJmNGNmMDhmMjdjOTQxYjIyMTE1NjIiLCJwIjoiaiJ9)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by levadadenys** [comment]
Button changes will be moved to https://github.com/code-dot-org/code-dot-org/pull/52117 once that PR is merged

File: apps/src/templates/CheckboxDropdown.jsx [hunk 1]
```diff
@@ -12,63 +18,89 @@ const CheckboxDropdown = ({
...
+        aria-label={`${name} filter dropdown`}
+      >
+        {checkedOptions.length > 0 && (
+          <FontAwesome
+            id={'check-icon'}
+            icon="check-circle"
+            title={i18n.filterCheckIconTitle({filter_label: label})}
+          />
+        )}
+        <Typography semanticTag="span" visualAppearance="body-one">
```
Comment (megcrenshaw) [inline apps/src/templates/CheckboxDropdown.jsx:55]: Really cool to see the Typography component coming to life with these! It's very clear what's going on and makes styling a breeze!

Comment (megcrenshaw) [reply apps/src/templates/CheckboxDropdown.jsx:55]: Thanks for your thoughtfulness with this.

File: apps/src/templates/CheckboxDropdown.jsx [hunk 2]
```diff
@@ -12,63 +18,89 @@ const CheckboxDropdown = ({
...
+              </label>
+            </li>
+          ))}
+          <div className={style.bottomButtonsContainer}>
+            <Button
+              id={'select-all'}
+              className={style.affectAllButton}
+              text={i18n.selectAll()}
+              onClick={onSelectAll}
+              styleAsText
```
Comment (megcrenshaw) [inline apps/src/templates/CheckboxDropdown.jsx:88]: TIL about this prop!

File: apps/src/templates/checkbox-dropdown.module.scss [hunk 3]
```diff
@@ -1,10 +1,100 @@
...
+    //    margin-right: 10px;
+    //    line-height: 20px;
+    //    text-align: center;
+    //    box-shadow: inset 0 1px 1px rgb(0 0 0 / 0.075);
+    //    content: "\f00c";
+    //    background: red;
+    //    border: 2px solid blue;
+    //    border-radius: 4px;
+    //  }
+    //}
```
Comment (megcrenshaw) [inline apps/src/templates/checkbox-dropdown.module.scss]: Nit: could you delete this before merging?

File: apps/style/code-studio/curriculum_catalog_container.module.scss [hunk 4]
```diff
@@ -2,11 +2,18 @@
...
   max-width: 60em;
   margin: 1em auto;
+
+  button {
+    margin: 0;
+  }
 }
 
-.catalogClearFiltersButton {
+button.catalogClearFiltersButton {
```
Comment (megcrenshaw) [inline apps/style/code-studio/curriculum_catalog_container.module.scss:15]: I'm not seeing where we use the `button` tag anymore. Are these styles being applied?

File: apps/src/templates/curriculumCatalog/CurriculumCatalog.jsx [hunk 5]
```diff
@@ -290,14 +294,14 @@ const CurriculumCatalog = ({curriculaData, isEnglish}) => {
             handleClearAll={() => handleClearAllOfFilter(filterKey)}
           />
         ))}
-        <button
+        <Button
           id="clear-filters"
-          type="button"
           className={style.catalogClearFiltersButton}
```
Comment (megcrenshaw) [inline apps/src/templates/curriculumCatalog/CurriculumCatalog.jsx:299]: Same question here –– not convinced these styles are doing anything at this point

File: apps/style/code-studio/curriculum_catalog_container.module.scss (hunk 4 above)
Comment (megcrenshaw) [reply apps/style/code-studio/curriculum_catalog_container.module.scss:15]: Oh I see –– the Button component uses the `button` tag. Ignore me!

**Review by megcrenshaw — COMMENTED** [review]
This is looking great, and I'm loving the coding patterns! I am worried that the `preventDefault()` methods are cascading into some unintended keyboard behavior. In particular, pressing "enter" when on a checkbox makes *all* the checkboxes selected, we aren't able to hit "enter" to trigger "clear filters" (but the "space" key does clear filters), and the "esc" key opens the dropdown.

It's a bit hard to understand what's happening in the video because you can't see the keys I'm pressing, but here's the video:
https://github.com/code-dot-org/code-dot-org/assets/9142121/9c03a49f-0ef9-442a-9cbb-9fc2923673d1

File: apps/src/templates/checkbox-dropdown.module.scss (hunk 3 above)
Comment (levadadenys) [reply apps/src/templates/checkbox-dropdown.module.scss]: yep, can be deleted already (checkbox is ready in different PR)

**Comment by levadadenys** [comment]
> This is looking great, and I'm loving the coding patterns! I am worried that the `preventDefault()` methods are cascading into some unintended keyboard behavior. In particular, pressing "enter" when on a checkbox makes _all_ the checkboxes selected, we aren't able to hit "enter" to trigger "clear filters" (but the "space" key does clear filters), and the "esc" key opens the dropdown.
> 
> It's a bit hard to understand what's happening in the video because you can't see the keys I'm pressing, but here's the video: https://github.com/code-dot-org/code-dot-org/assets/9142121/9c03a49f-0ef9-442a-9cbb-9fc2923673d1

Yeah, but it's not because of preventDefault. The reason for space/enter issue is that our buttons didn't have the type prop on it, and so they had submit type by default and that's how button type submit behaves in currents structure.
Since button updates from https://github.com/code-dot-org/code-dot-org/pull/52117 are merged, I've fixed the issue with enter/space bar. Now the checkbox will be updated once you click space (default input checkbox behavior) and all the buttons can be clicked via space or enter.

The escape logic for main dropdown button - not sure if it's a bug or default behavior, but I suggest we look deeper on it in separate PR since the idea of this PR is to give basic styling updates and not to block @code-dot-org/acquisition-products-team in case something won't from design system won't be ready on time.

**Comment by levadadenys** [comment]
^^^ fixed issues from comments in last commit

**Comment by megcrenshaw** [comment]
The space and enter keys are working as expected on the dropdowns! The one exception is the "Clear filters" button –– that should work with the "enter" key but it's only working with the "space" key (which is different behavior from what's in production).

Agreed that the weird escape behavior is a leftover bug from before, and no need to address that in your styling work. @TurnerRiley is it on your radar that the "escape" key is opening dropdowns? It should be that the "escape" keys only closes dropdowns.

**Review by megcrenshaw — APPROVED** [review]
Merging this as-is introduces a bug not present earlier: that the "Clear filters" button doesn't get triggered on pressing "enter." It does sound like this PR is blocking other work, so I'm going to approve it.

Two tickets are added to our backlog:
1) https://codedotorg.atlassian.net/browse/ACQ-637 (escape key on dropdowns)
2) https://codedotorg.atlassian.net/browse/ACQ-636 (enter key on clear filters button)
It's unclear to me who owns the second task, but the first task will definitely be tackled by someone on our team!

This looks great visually –– thanks for your thoughtful work on this.

**Comment by breville** [comment]
Nice to see these improvements.  Some quick thoughts from playing with it:

- While each row has a grey hover highlight if you are anywhere over the row, it's only clickable when over the text or checkbox part of the row.  It would be good to make the entire row clickable, and for the cursor to reflect that.
- When toggling an option, it's easy to do a double-click, leaving some dropdown text highlighted.  It would be good to disable selection of the text.
- When any option is selected in a dropdown, a checkmark appears at the top, shifting the following dropdowns.  It might be good to create space for that checkmark at all times so that the buttons don't shift around.
- It might be desirable for touch target sizing, but each row seems quite big, and for the longer lists of dropdowns, some users will have to scroll their window up and down to see all options.  It might be worth shrinking everything a bit.
- A smaller detail, but the baseline of the "Filter by" text doesn't match the baseline of the dropdown texts.
- Should the down caret used here match the down caret used elsewhere, for example in the header?

**Comment by breville** [comment]
Some additional notes:

- We don't appear to be using HTML `select` tags for dropdowns.  This means we don't get the convenience of browser-provided dropdowns, and on mobile devices in particular we don't get their special touch-friendly implementations.  
  - Do we prefer some benefits from a custom implementation?
  - Are there accessibility impacts because of this?
  - See https://code.org/learn for examples of using `select` tags for dropdowns.
- On narrower browser widths for mobile devices, the buttons get a bit messy
  - For example, the checkmark is positioned inconsistently and the buttons end up different sizes.

Screenshot 2023-06-09 at 8 17 45 AM

**Comment by megcrenshaw** [comment]
Thank you, Brendan, for looking at this! So appreciate it. I agree that we should be using HTML `select` tags for the dropdowns. We are splitting this up a bit strangely because Denys is focused on CSS styling (which so often relies on the HTML structure), but I'm thinking this split of work between Denys and the rest of our team as follows, keeping in mind that the `select` tags for dropdowns might fix some of these things right away:

- Make entire row clickable, with cursor to reflect that: Our team
- Disable selection of text: Denys
- Space for checkmark with text: Denys
- Size of row: Denys
- Filter by text baseline: Denys
- Down caret use: @dmcavoy should our arrows be consistent with what's in the header? That would fall on our team, but it looks like it's just a different FontAwesome icon. Also, do we want the arrow direction to change based on if the dropdown is collapsed or not?
- Select tags: Our team
- Mobile: @dmcavoy can you chime in on how important mobile is right now? I was focused on making the cards "doable" on mobile, but maybe not perfect. I do think mobile styling would fall on Denys if possible.

@TurnerRiley and @levadadenys I'm thinking it makes sense for our team to do the HTML work first with the select tags and the clickability of the whole row. And then Denys can follow up with the styling adjustments. How does that sound? How does the split of the work seem to you both?

**Comment by dmcavoy** [comment]
The down caret is probably a design team question. Meg can you ask in the #design channel? 

For mobile we are not prioritizing making the mobile perfect right now just doable.

**Comment by megcrenshaw** [comment]
Asked in the design channel here: https://codedotorg.slack.com/archives/C0SUN2SSF/p1686330140352379

One more thing I noticed: the focus state on the dropdowns doesn't have high enough contrast for accessibility. That's a design thing –– @TurnerRiley can you include that in the Jira ticket(s) you're writing as well?
image

**Comment by levadadenys** [comment]
Wow guys, the pr was merged a while ago and I've just seen all the comments appeared here after merge. Let me try answer all the questions one by one:
@breville 
* Fixed in the [second PR](https://github.com/code-dot-org/code-dot-org/pull/52251) - `While each row has a grey hover highlight if you are anywhere over the row, it's only clickable when over the text or checkbox part of the row. It would be good to make the entire row clickable, and for the cursor to reflect that.`
* Nice catch, can be fixed with couple lines of css, thank you! - `When toggling an option, it's easy to do a double-click, leaving some dropdown text highlighted. It would be good to disable selection of the text.`
* Nice catch, should be discussed with the designers - `When any option is selected in a dropdown, a checkmark appears at the top, shifting the following dropdowns. It might be good to create space for that checkmark at all times so that the buttons don't shift around.`
* I believe it's also fixed or partially fixed into [second PR](https://github.com/code-dot-org/code-dot-org/pull/52251). Changing more then that should be discussed with designers - `It might be desirable for touch target sizing, but each row seems quite big, and for the longer lists of dropdowns, some users will have to scroll their window up and down to see all options. It might be worth shrinking everything a bit.`
* Nice catch, should be fixed with 1 line of css. `A smaller detail, but the baseline of the "Filter by" text doesn't match the baseline of the dropdown texts.`
* This dropdown is part of new Design System (DSCO) so the answer is no, we have other icons for new dropdown - `Should the down caret used here match the down caret used elsewhere, for example in the header?`

**Comment by levadadenys** [comment]
@breville 
* Oh, that's a great question. Yeah, I guess select will be a better solution if we can implement all the custom styles we need from the designs - `We don't appear to be using HTML select tags for dropdowns. This means we don't get the convenience of browser-provided dropdowns, and on mobile devices in particular we don't get their special touch-friendly implementations.`

* Nice catch as well, need to discuss it with design team -`On narrower browser widths for mobile devices, the buttons get a bit messy`

**Comment by levadadenys** [comment]
@megcrenshaw I'm focused not on css only, but on creating reusable Design System across all the apps, meaning creating reusable components with new styles from design system. 
I'm sorry for inconveniences that was brought by it, but the only reason of it is that the dropdown component was not ready by the time you got to this functionality. (The initial idea here was that once you come to this page you simply use it from `/componentsLibrary` or `dsco_`)

That being said, I believe all the work re dropdown component should be made by me while you and your team keep focusing on the page functionality. The only thing that we've already discussed with @TurnerRiley (it was before I've seen all of these comments) is that he wanted to make dropdown close on escape key press. (But actually, I can make that functionality as well since that's the part of dropDown component).

* Some of the changes are already done by me, some will be done in another PR right after I finish answering questions in this pr. All the list should be done by me I think - `Make entire row clickable, with cursor to reflect that: Our team
Disable selection of text: Denys
Space for checkmark with text: Denys
Size of row: Denys
Filter by text baseline: Denys
Down caret use: @dmcavoy should our arrows be consistent with what's in the header? That would fall on our team, but it looks like it's just a different FontAwesome icon. Also, do we want the arrow direction to change based on if the dropdown is collapsed or not?
Select tags: Our team
Mobile: @dmcavoy can you chime in on how important mobile is right now? I was focused on making the cards "doable" on mobile, but maybe not perfect. I do think mobile styling would fall on Denys if possible.`

**Comment by levadadenys** [comment]
@megcrenshaw 
* Nice catch, thank you! That's again as with checkbox my mistake (honestly don't know how it happened). Will fix it. - `One more thing I noticed: the focus state on the dropdowns doesn't have high enough contrast for accessibility. That's a design thing –– @TurnerRiley can you include that in the Jira ticket(s) you're writing as well?`

**Comment by levadadenys** [comment]
@breville @megcrenshaw @dmcavoy 
Thanks a lot for your attention and notes! Sorry for long response. Hope I answered all of your questions now.
Please, ping me in slack if it takes me more then 1 day to respond in github or you're commenting merged PR.

**Comment by levadadenys** [comment]
@megcrenshaw also, to double check and clarify:
the issue when filter by section has more width then the screensize should be done by @code-dot-org/acquisition-products-team once designs for that are ready, cause that's part of page functionality, not design system (unless it'll be part of it by that time/will use some page/section template from DSCO).
Making those dropdown buttons look as expected is however my responsibility, so I'll check on how to fix the checkmark is positioned inconsistently and the buttons end up different sizes. (fixed in [pr](https://github.com/code-dot-org/code-dot-org/pull/52318))

**Comment by megcrenshaw** [comment]
That sounds good to me, thanks for the clarifying, Denys!

**Comment by TurnerRiley** [comment]
> Asked in the design channel here: https://codedotorg.slack.com/archives/C0SUN2SSF/p1686330140352379
> 
> One more thing I noticed: the focus state on the dropdowns doesn't have high enough contrast for accessibility. That's a design thing –– @TurnerRiley can you include that in the Jira ticket(s) you're writing as well? image

On it! Thank you @megcrenshaw and @levadadenys for the work on this!

**Pull request merged**